### PR TITLE
fix(sha3)_: support hex string

### PR DESCRIPTION
**Pull request by qfrank**
This PR adds a check to determine if a `str` starts with `0x`.

Relate mobile [issue](https://github.com/status-im/status-mobile/issues/21819).

This is what `web3` implemented:
```ts
export const sha3 = (data: Bytes): string | undefined => {
    let updatedData: Uint8Array;

    if (typeof data === 'string') {
        if (data.startsWith('0x') && isHexStrict(data)) {
            updatedData = hexToBytes(data);
        } else {
            updatedData = utf8ToBytes(data);
        }
    } else {
        updatedData = data;
    }
    const hash = keccak256Wrapper(updatedData);

    // EIP-1052 if hash is equal to c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470, keccak was given empty data
    return hash === SHA3_EMPTY_BYTES ? undefined : hash;
};
```

### CODE REVIEW

**Review by flexsurfer — APPROVED** [review]

**Comment by anastasiyaig** [comment]
@saledjenic pls take a look

File: abi-spec/utils_test.go [hunk 1]
```diff
@@ -36,6 +36,9 @@ func TestNumberToHex(t *testing.T) {
 
 func TestSha3(t *testing.T) {
 	require.Equal(t, "48bed44d1bcd124a28c27f343a817e5f5243190d3c52bf347daf876de1dbbf77", Sha3("abcd"))
+	require.Equal(t, "79bc958fa37445ba1b909c34a73cecfa4966a6e6783f90863dd6df5a80f96ad0", Sha3("12786e7b2111caae36dd91aca91ce627f26fa3c77018a98880ab50a82ac6b6aa835f6bbf96da54aa6b88ceffb8107ef40a4ef8a85bf4eb0e81a9464e0a27fcf3"))
+	require.Equal(t, "04d874cdee68658bd64a07b6180dd36b735b60876ca79a29f88114bc78e6fc32", Sha3("0x12786e7b2111caae36dd91aca91ce627f26fa3c77018a98880ab50a82ac6b6aa835f6bbf96da54aa6b88ceffb8107ef40a4ef8a85bf4eb0e81a9464e0a27fcf3"))
+	require.Equal(t, "04d874cdee68658bd64a07b6180dd36b735b60876ca79a29f88114bc78e6fc32", Sha3("0x12786E7b2111caae36dd91aca91ce627f26fa3c77018a98880ab50a82ac6b6aa835f6bbf96da54aa6b88ceffb8107ef40a4ef8a85bf4eb0e81a9464e0a27fcf3"))
```
Comment (igor-sirotin) [inline abi-spec/utils_test.go:41]: These 2 lines are completely equal 🤔

**Review by saledjenic — APPROVED** [review]

**Review by igor-sirotin — COMMENTED** [review]
@qfrank I'm ok to merge this for the release.

But for `develop` branch please fix the code duplication. We already have similar functions used in this file:
https://github.com/status-im/status-go/blob/3cf8b9b67457633655708f848290f92c67ba1bfd/abi-spec/utils.go#L149-L150

Perhaps `Sha3` should do these 2 conversions.

**Review by igor-sirotin — APPROVED** [review]

File: abi-spec/utils_test.go (hunk 1 above)
Comment (saledjenic) [reply abi-spec/utils_test.go:41]: They are not, in the second you have hex with some digits in uppercase format.

**Review by ilmotta — APPROVED** [review]

File: abi-spec/utils_test.go (hunk 1 above)
Comment (igor-sirotin) [reply abi-spec/utils_test.go:41]: oh indeed, pardon

File: abi-spec/utils.go [hunk 2]
```diff
@@ -20,6 +24,8 @@ var addressBasicPattern = regexp.MustCompile("(?i)^(0x)?[0-9a-f]{40}$")
 var addressLowerCasePattern = regexp.MustCompile("^(0x|0X)?[0-9a-f]{40}$")
 var addressUpperCasePattern = regexp.MustCompile("^(0x|0X)?[0-9A-F]{40}$")
 
+var hexStrictPattern = regexp.MustCompile(`(?i)^((-)?0x[0-9a-f]+|(0x))$`)
```
Comment (richard-ramos) [inline abi-spec/utils.go:27]: What does the `(-)?` represent here?

Comment (qfrank) [reply abi-spec/utils.go:27]: refer [here](https://github.com/web3/web3.js/blob/4.x/packages/web3-validator/src/validation/string.ts#L26) , just wanted to align it with `web3` library

**Comment by qfrank** [comment]
> @qfrank I'm ok to merge this for the release.
> 
> But for `develop` branch please fix the code duplication. We already have similar functions used in this file:
> 
> https://github.com/status-im/status-go/blob/3cf8b9b67457633655708f848290f92c67ba1bfd/abi-spec/utils.go#L149-L150
> 
> Perhaps `Sha3` should do these 2 conversions.

There was a comment "implementation referenced from https://github.com/ChainSafe/web3.js/blob/edcd215bf657a4bba62fabaafd08e6e70040976e/packages/web3-utils/src/utils.js#L107" @igor-sirotin , its purpose was to align with `web3`

File: abi-spec/utils.go (hunk 2 above)
Comment (igor-sirotin) [reply abi-spec/utils.go:27]: cargo-culting, sir? 😄

**Comment by igor-sirotin** [comment]
> its purpose was to align with `web3` 

Sure, and I even found that code and checked it. But the fact that code int that library is polluted, doesn't mean we should do the same to our codebase 🙂

Less duplication -> Less chance to face a bug later

**Pull request merged**